pkg/distributor/distributor_test.go

```diff
@@ -823,25 +838,36 @@ func prepare(t *testing.T, numIngesters, happyIngesters int, queryDelay time.Dur
 		})
 	}
 
-	// Mock the ingesters ring
-	ingesterDescs := []ring.IngesterDesc{}
+	// Use a real ring with a mock KV store to test ring RF logic.
+	ingesterDescs := map[string]ring.IngesterDesc{}
 	ingestersByAddr := map[string]*mockIngester{}
 	for i := range ingesters {
 		addr := fmt.Sprintf("%d", i)
-		ingesterDescs = append(ingesterDescs, ring.IngesterDesc{
+		ingesterDescs[addr] = ring.IngesterDesc{
 			Addr:      addr,
+			Zone:      addr,
+			State:     ring.ACTIVE,
 			Timestamp: time.Now().Unix(),
-		})
+			Tokens:    []uint32{uint32((math.MaxUint32 / numIngesters) * i)},
+		}
 		ingestersByAddr[addr] = &ingesters[i]
 	}
 
-	ingestersRing := mockRing{
-		Counter: prometheus.NewCounter(prometheus.CounterOpts{
-			Name: "foo",
-		}),
-		ingesters:         ingesterDescs,
-		replicationFactor: 3,
-	}
+	store := consul.NewInMemoryClient(ring.GetCodec())
+	err := store.Put(context.Background(), ring.IngesterRingKey, &ring.Desc{
+		Ingesters: ingesterDescs,
+	})
+	require.NoError(t, err)
+
+	ingestersRing, err := ring.New(ring.Config{
+		KVStore: kv.Config{
+			Mock: store,
+		},
+		HeartbeatTimeout:  60 * time.Minute,
+		ReplicationFactor: 3,
+	}, ring.IngesterRingKey, ring.IngesterRingKey)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), ingestersRing))
 
```

Comment:
Once the ring has been started, it's not guaranteed that the initial sync has been completed (the local ring state may still be empty). To avoid flaky tests you should add this:

```suggestion
	// Wait until the ring client has updated its status to avoid flaky tests.
	test.Poll(t, time.Second, numIngesters, func() interface{} {
		return ingestersRing.IngesterCount()
	})
```